Add matchmaking functionality

shmashing opened the pull request:
This PR adds in the logic to match up teams based on their ELO. The method I chose is the simulated annealing approach. Essentially, we randomly assign each guardian to a team. Then we start iterating and randomly swapping two guardians on two different teams. Then, we measure the ELO difference of those two teams from before and after the swap and take the difference. If it is less than zero, that means the swap is favorable and we are closer to having equal ELOs. Otherwise, we roll a random probability and compare it to the change in ELO. If the rolled probability is lower, then we will still keep the assignment and keep moving. This allows for randomness in the way the teams settle to let it reach a more equal state

shmashing committed:
Add in Base Models

This commit adds in the base models that will be used for the web app

shmashing committed:
Add Average Eff To Guardian

This commit adds the average efficiency field to the guardian object because i think this will be a field we will want to track over time

shmashing committed:
Averages Usually Arent Ints

This commit changes the average efficiency field from an int to double because why would that ever be an int

shmashing committed:
Break Out Model Concerns

This commit breaks out different guardian conerns from the guardian class.

A guardians elo will now be represented by their own table so we can better track elo changes over times.
A guardians efficiency will also now be tracked by its own table for the same reasons.

This commit also changes the structure of the match object so it is not limited by two teams.

shmashing committed:
Add Models Folder Back to csproj

shmashing committed:
Add In Base CRUD For Guardians

This commit adds in the base file structure and code to be able to get all guardians, get a specific guardian, create a guardian, and also delete them.

This commit also adds in the logic to update a guardians total kills and total deaths

shmashing committed:
Remove Old Migration Logic

This commit removes leftover code from how I was thinking about handling migrations before.

Now we will just use entity framework to handle all that jazz

shmashing committed:
Add Get Single Guardian Endpoint

This commit adds in the logix to get a single guardian from the DB.

shmashing committed:
Update DB Models and Migrations

shmashing committed:
Merge branch 'master' into add-base-models

shmashing committed:
Add in the updated DB context object. Originally we were planning on using multiple context objects for our models but that seems like too much overhead. It is now simplified to use a single context

shmashing committed:
Merge branch 'add-base-models' of github.com:shmashing/destiny-scrim-tracker into add-base-models

shmashing committed:
Add ID Prefixes as static classes to make sure they remain consistent and if we need to change them we only have to change them in a single place

shmashing committed:
Add Front To Back Controls For Matches

This commit adds the front to back workflow for adding matches, updating matches, deleting matches, and all that goodness. This commit also adds in the workflow for updating the postgres database on startup

shmashing committed:
Remove Unused Using Directives

shmashing committed:
Delete Old Migration Files

This commit removes some broken migration files. I was using a database context class that is no longer used. It should get deleted but these migration files and snapshots depend on it

shmashing committed:
Add EF Core Dependencies

shmashing committed:
Add Initial Migrations Files

shmashing committed:
Add Dotnet EF Tool To API Package

shmashing committed:
Downgrade EF CLI Version

shmashing committed:
Add Connection URI Parser

Turns out, Heroku provides the connection to the Postgres db as a URI. C# needs it to be a connection string. This commit adds in a little bit of parsing to get out the necessary information into a form we can use

shmashing committed:
Merge branch 'master' of github.com:shmashing/destiny-scrim-tracker into add-frontend-views-and-functionality

shmashing committed:
Add Quick Nav Links

This commit adds in nav links so that we can navigate through the UI to each of the pages.

shmashing committed:
Add In Formatting to Efficiency

This commit adds in some formatting to the efficiency on the main page, and also adds nav back to the home page from the all guardians page

shmashing committed:
Add In Auth Middleware

This commit adds in middleware to the API that whitelists IPs so we can block requests from anyone whos not on the Admin list

shmashing committed:
Remove Access to All Endpoints

shmashing committed:
Add My IP To WHitelist

shmashing committed:
Remove IP Whitelist from Source

shmashing committed:
Change Where Looking Admin Safe List Env

shmashing committed:
Add Debug Logging Statements To Middleware Cos Its No Working

shmashing committed:
Rework IP Logic Cos It STILL No Work :(

shmashing committed:
Still Trying :(

shmashing committed:
More Logging :(

shmashing committed:
Change Up the Logic Lil Bit

shmashing committed:
Try More Logix CoS ITS STILL NOT WORKING

shmashing committed:
ZOMG It Works

This commit removes some of the debug log statements that aren't necessary anymore

shmashing committed:
Remove Auth Guards

shmashing committed:
FaceLift With Styles

shmashing committed:
Remove Client Secret From Source

shmashing committed:
Add IN Auth Guards To All Non Get Routes

shmashing committed:
More FaceLifty Types of Things

shmashing committed:
Fix Styling Bug And Add Minified CSS

shmashing committed:
Last Commit On This PR

This commit adds in Guardian Performance tracking over time via graphs and charts and is hella dope

shmashing committed:
Yep Forgot To commit A Buncho Stuff. WHoops

shmashing committed:
Fix Divide by Zero Bug

shmashing committed:
Update Bundler

shmashing committed:
Update BundleConfig

shmashing committed:
Fix Production Layer Script Importing

shmashing committed:
Add Icons

shmashing committed:
Actually Commit Files

shmashing committed:
Add Delete Button For Matches

shmashing committed:
Add Delete Guardian Functionality to UI

shmashing committed:
Fix Guardian Display Bug and Table Overlap Styles

shmashing committed:
Fix BadBadBug

This commit fixes a bug that would set efficiency to zero if a guardian went flawless in a match. That would be very much no bueno

shmashing committed:
Add Sorting To Guardians Grid

shmashing committed:
Fix Row Ordering

shmashing committed:
Adds The Match Making Algorithm

This commit adds in the logic to match up teams based on their ELO. The method I chose is the simulated annealing approach. Essentially, we randomly assign each guardian to a team. Then we start iterating and randomly swapping two guardians on two different teams. Then, we measure the ELO difference of those two teams from before and after the swap and take the difference. If it is less than zero, that means the swap is favorable and we are closer to having equal ELOs. Otherwise, we roll a random probability and compare it to the change in ELO. If the rolled probability is lower, then we will still keep the assignment and keep moving. This allows for randomness in the way the teams settle to let it reach a more equal state

shmashing committed:
Remove Authorize Attribute For MatchMaking

This commit removes the requirement for a login to run the matchmaking.

benprime commented on Destiny.ScrimTracker.Logic/Repositories/GuardianEfficiencyRepository.cs:
```diff
@@ -0,0 +1,104 @@
+        public void UpdateGuardianEfficiency(GuardianEfficiency guardianEfficiency)
+        {
+            _databaseContext.Add(guardianEfficiency);
+            _databaseContext.SaveChanges();
```
Might be a good idea to use all the async methods in EF: SaveChangesAsync(), ToListAsync(), FirstOrDefaultAsync, etc.

shmashing committed:
Rename API to App

shmashing committed:
Add Some Styling to the Matchmaking Page

shmashing committed:
Add List of Guardians To Scrollable Tbale and Add Counter To Keep Track

shmashing committed:
Add A Bit Of Styles To MatchMaking

shmashing committed:
Rework Create Match Form to Make it Better

This commit also removes the probability factor in the matchmaking algorithm for now. It was skewing the matchmaking results with its randomness, and needs to be dialed in a bit more

shmashing committed:
Clean up Create Match Page

shmashing committed:
Clean Up Error Handling

This commit cleans up the error handling on matchmaking and actually surfaces errors back to the user

shmashing committed:
fix bug that i commited

jaredonline commented on Destiny.ScrimTracker.App/Controllers/MatchController.cs:
```diff
@@ -32,55 +32,39 @@ public IActionResult Index()
+            Console.WriteLine($"Key count: {formKeys.Count()}");
             var teams = new List<MatchTeam>();
-            for (var i = 0; i < numOfTeams; i++)
+            for (var i = 0; i < 2; i++)
```
Why hardcode this `2`?

shmashing replied:
Couldn't figure out how to make the form reactive enough to be able to add team forms on the fly. Added an issue to track it. For now, since two teams is our primary use case, i figured this would do?

jaredonline commented on Destiny.ScrimTracker.Logic/Services/MatchMakingService.cs:
```diff
@@ -0,0 +1,254 @@
+    public class MatchMakingService : IMatchMakingService
+    {
+        private const double _epsilon = 0.10;
+        private const double _alpha = 0.999;
```
Comments about what these do and what they control would be good

jaredonline commented on Destiny.ScrimTracker.Logic/Services/MatchMakingService.cs:
```diff
@@ -0,0 +1,254 @@
+        IEnumerable<MatchMadeTeam> MatchTeams(IEnumerable<string> guardiansId, int numberOfTeams);
+    }
+    
+    public class MatchMakingService : IMatchMakingService
```
I'd add a comment to this class explaining where you got this annealing function from and how you applied it

jaredonline commented on Destiny.ScrimTracker.Logic/Services/MatchMakingService.cs:
```diff
@@ -0,0 +1,254 @@
+                if (delta < 0)
+                {
+                    teams = KeepNextConfiguration(nextTeamConfiguration);
+                }
```
Without doing the annealing part below you're essentially doing a random hill climb, which will find a local maxima but might miss a global maxima. I'm not an expert in matchmaking algorithms but I think having a less accurate global maxima is the preferred solution.

jaredonline commented on Destiny.ScrimTracker.Logic/Services/MatchMakingService.cs:
```diff
@@ -0,0 +1,254 @@
+//                    {
+//                        teams = KeepNextConfiguration(nextTeamConfiguration);
+//                    }
+//                }
```
I'm trying to reason through the probability function you're got here and the implications it has....

`Math.Exp(x)` is `e^x`, which approaches zero for negative values of `x`, is 1 when `x` is 0 and increases exponentially when `x` is positive.

If `x` is positive, then `delta` is negative, which means its less than `0` which means we won't hit that case.
If `x` is zero, then `Math.Exp` is 1, so we will always keep that case. This is interesting and probably not what we want, but also unlikely.

That leaves cases where `delta` is positive, which will make `x` negative. The probability will be close to 1 when `x` is barely negative, for example when `delta` is very small. When `0 > x > -1`, (ie when `delta` is smaller than `temperature`), the probability will be extremely close to 1. When `delta` is large, the probability will be smaller. This means we favor small `delta` changes, and our tolerance for what is small gets smaller as the temperature decreases.

_If_ most changes result in _worse_ configurations (meaning a positive `delta`), _and_ most configurations are relatively similar to one another (because ELOs are similar), then it would seem that this probability will almost always result in us keeping the "worse" configuration.  If by chance, the algorithm comes up with a better solution, we'll use that until a worse one is encountered, and then we'll use the worse one. In essence we'll likely walk away from a good solution until we hit some point where a random change is just as likely to be bad as it is good, then we'll oscillate between moving towards a good solution and a bad solution. We'll do this until `temperature` is very small, then we'll only start moving towards good solutions. The thing is that by the time `temperature` is small enough, we don't have very many iterations left in the algorithm, we're too "cool" essentially to walk our way back to a global maximum. 

That's my theory anyway. You may just do something like `Math.Exp(-1000 * delta / temperature)`, or start `temperature = 4` and `_alpha = .00999`.

shmashing replied:
Yea.. I just need to sit down and dig into the math and modeling here, I think. I also think you're thinking about this in an inverted kinda way. Delta is guaranteed to be > 0 in this case. If it's not, then we don't even hit the e^x probability function.

jaredonline replied:
> That leaves cases where `delta` is positive, which will make `x` negative

That's the case I talk about the most. It could also be `0`, which I think is unlikely, and would be weird.

shmashing replied:
oh right i can read.. whoooops 😂 i did try the suggestions you left at the bottom there but they were so small they were rounded to 0 lol and yea, i don't think we need to start with temp = 400, just overkill but i dunno

jaredonline replied:
Maybe then leave temp and alpha alone and try multiplying Delta by 1000.

You could do some data gathering by logging out all the values used on every iteration along with a) did we use the probability branch and b) on that branch did we actually take the step backward.

Non-deterministic functions are tough to debug, but logging out lots of information can help spot weirdness.

shmashing committed:
Merge remote-tracking branch 'origin/master' into add-matchmaking-functionality

shmashing committed:
Fix Regression Introduced by Merge

shmashing committed:
Fixing More Merge Regressions :(

shmashing committed:
Add Comments To Matchmaking Service

To give credit to where our solution was derived from, and help explain the decisions that were made
